content/docs/quickstart/go.md

```diff
@@ -3,98 +3,150 @@ title: Go Quick Start
+          ```sh
+          $ unzip protoc-{{< param protoc-version >}}-linux-x86_64.zip -d $HOME/local
+          ```
 
-Change to the example directory
+       3. Update your environment's path variable to include the path to the
+          `protoc` executable. For example:
 
-```sh
-$ cd $GOPATH/src/google.golang.org/grpc/examples/helloworld
-```
+          ```sh
+          $ export PATH="$PATH:$HOME/local/bin"
+          ```
 
-gRPC services are defined in a `.proto` file, which is used to generate a
-corresponding `.pb.go` file. The `.pb.go` file is generated by compiling the
-`.proto` file using the protocol compiler: `protoc`.
+ 2. The `protoc` plugin for Go (`protoc-gen-go`) was installed as a dependency
+    of the `grpc` module. You can confirm this, or install the plugin, using the
+    following command:
 
-For the purpose of this example, the `helloworld.pb.go` file has already been
-generated (by compiling `helloworld.proto`), and can be found in this directory:
-`$GOPATH/src/google.golang.org/grpc/examples/helloworld/helloworld`
+    ```sh
+    $ go get -u github.com/golang/protobuf/protoc-gen-go
+    ```
 
-This `helloworld.pb.go` file contains:
+ 3. Update your `PATH` so that the `protoc` compiler can find the plugin:
 
-- Generated client and server code.
-- Code for populating, serializing, and retrieving our `HelloRequest` and
-  `HelloReply` message types.
+    ```sh
+    $ export PATH="$PATH:$(go env GOPATH)/bin"
+    ```
 
-### Try it!
+### Copy the example
 
-To compile and run the server and client code, the `go run` command can be used.
-In the examples directory:
+The example code is part of the `grpc` source, which you fetched by following
+steps of the previous section. You'll need a local copy of the example code to
+work through this quick start.
 
-```sh
-$ go run greeter_server/main.go
-```
+ 1. Choose a suitable working directory, and ensure that it exists:
 
-From a different terminal:
+     ```sh
+    $ export MY_EXAMPLES="$HOME/examples"
+    $ mkdir -p "$MY_EXAMPLES"
+    ```
 
-```sh
-$ go run greeter_client/main.go
-```
+ 2. Copy the example source:
+
+    ```sh
+    $ EX_SRC_DIR="$(go env GOPATH)/pkg/mod/google.golang.org/grpc@{{< param grpc_release_tag >}}/examples"
+    $ cp -R "$EX_SRC_DIR/helloworld" "$MY_EXAMPLES"
+    ```
+
+ 3. Ensure that the example files are writable since you'll be making changes soon:
+
+    ```sh
+    $ chmod -R u+w "$EX_SRC_DIR/helloworld"
+    ```
+
+ 4. Change to the example's directory:
+
+    ```sh
+    $ cd "$MY_EXAMPLES/helloworld"
+    ```
+
+ 5. Setup the example as a module:
 
-If things go smoothly, you will see the `Greeting: Hello world` in the client side output.
+    ```sh
+    $ go mod init examples/helloworld
+    ```
+
+ 6. Adjust import paths to reference local packages (rather than those from the
+    original `google.golang.org/grpc` module):
+
+    ```sh
+    $ perl -pi -e 's|google.golang.org/grpc/||g' greeter_{client,server}/main.go
+    ```
```

Comment:
@dfawley - it would be great if the original example sources were already setup in this way (so that we can avoid these last two steps). Might that be possible?

Reply:
I missed this comment somehow.  So make the import paths relative?  I think that would be OK if it's not considered an anti-pattern (we always use absolute paths).

Reply:
My concern is that having the reader follow 6 steps _just_ to copy an example's source seems like a lot of work -- especially for a _quick_ start :). Is there any way that we can avoid some of these 6 steps?

My original suggestion what that the `helloworld` example be defined as a separate module already (named, say, `examples/helloworld`) in the `grpc-go` repo, but I don't know if it makes sense to do so (that is, to define a module inside of a module).

Reply:
To be fair, one whole step is changing directories.

Is there a curl command we can use to fetch the examples from github?  That would avoid the need to chmod them.

The "Download" link on any directory in github links you to something like this: https://download-directory.github.io/?url=https://github.com/grpc/grpc-go/tree/master/examples/helloworld

But then it asks you for a token. :unamused:

> My original suggestion what that the helloworld example be defined as a separate module already (named, say, examples/helloworld) in the grpc-go repo, but I don't know if it makes sense to do so (that is, to define a module inside of a module).

Yes, it's totally OK to have modules inside other modules.  Such a change is a breaking change, though.  But this is just an example.  And there's some amount of shenanigans you may need to do to avoid the conflict of older versions of grpc-go having the same packages.  I think you do this by setting the sub-module to require a version of grpc-go _after_ (or at the same commit where) the sub-module was forked.  But I'm not able to find the right documentation right now (I'll keep searching).

Reply:
Of course, after 5 minutes of failed searches....then my first query after sending this turned up what I was looking for:

https://github.com/golang/go/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository

Reply:
Thanks for the link.

If you are in general agreement with the changes made in this PR, I'd like to get it merged before moving on to a possibly simpler solution. Would you agree?

Reply:
I'm generally fine with the contents of this PR.  One thing I dislike is the use of `perl`, here, which is not installed by default on all systems.  Can this be changed to use `sed` instead?

Reply:
On other projects I've encountered problems with the use use `sed -i` -- e.g., see this [SO post](https://stackoverflow.com/questions/9704020/recursive-search-and-replace-in-text-files-on-mac-and-linux). That's why I'm proposing the use of perl here.

(I've just tried it again and you must use `-i ''` under macOS and `-i` (w/o any argument) under Linux.

Reply:
(Hopefully the use of this command will only be temporary :)).

Reply:
Here's the PR for making `helloworld` a module: https://github.com/grpc/grpc-go/pull/3540

Reply:
If step 5 was changed to:

`go mod init google.golang.org/examples/helloworld`

Then would step 6 be unnecessary?

Reply:
Yes since you're altering the path to drop the "/grpc/" part. Without step 6 the code would refer to the `.pb.go` files from the official `grpc` module.

Reply:
"Yes" step 6 would be _unnecessary_?  Or "Yes" it would still be necessary?

If the former, we can just make that change and not do anything in the `grpc-go` repo (I think?).

Reply:
Oops, apologies for being unclear in my answer. Step 6 would still be necessary because `go mod init google.golang.org/examples/helloworld` essentially changes the path to the example. The quick start also (now) changes the path to the example.

Btw, did you mean to write: `go mod init google.golang.org/grpc/examples/helloworld` -- which would preserve the path to the example, merely making `helloworld` an independent module? In that case we run into the ambiguous import problem.

Might you want us to discuss this (well, https://github.com/grpc/grpc-go/pull/3540, actually) over a VC some time tomorrow?

Reply:
> Btw, did you mean to write: go mod init google.golang.org/grpc/examples/helloworld -- which would preserve the path to the example, merely making helloworld an independent module? In that case we run into the ambiguous import problem.

Yes, that's what I meant, sorry.  OK, I see we still would also need a `go.mod` inside the grpc repo so that it removes those paths from its module.

At this point, given all the constraints I can see, and the options you laid out in https://github.com/grpc/grpc-go/pull/3540, what I'd prefer is adding a single `go.mod` to `grpc/examples` (3), and having the quick start guide recommend copying all of `grpc/examples` elsewhere to work on.  If that SGTY then I don't think there is any need to meet.  If you would prefer another direction or just want to explore all/some of the options further, then I'm happy to meet and am available tomorrow.  Feel free to email me (or use hangouts) and we can set up a time.

Reply:
I answered in https://github.com/grpc/grpc-go/pull/3540#issuecomment-614973544 -- so that we could move the discussion back over there.